src/http/matcher/mod.rs

```diff
@@ -880,6 +696,517 @@ impl<State, Body> crate::service::Matcher<State, Request<Body>> for HttpMatcherK
+                HttpMatcher::custom(TrueMatcher),
+                true,
+            ),
+            (
+                HttpMatcher::custom(FalseMatcher),
+                HttpMatcher::custom(FalseMatcher),
+                HttpMatcher::custom(FalseMatcher),
+                true,
+            ),
+        ];
+
+        for (a, b, c, expected) in cases {
+            let matcher = a.negate().or(b).or(c);
+            let req = Request::builder().body(()).unwrap();
+            assert_eq!(
+                matcher.matches(None, &Context::default(), &req),
+                expected,
+                "({:#?}).matches({:#?})",
+                matcher,
+                req
+            );
+        }
+    }
+
+    #[test]
+    fn test_matcher_ors_combination_negated() {
+        let cases = [
+            (
+                HttpMatcher::custom(TrueMatcher),
+                HttpMatcher::custom(TrueMatcher),
+                HttpMatcher::custom(TrueMatcher),
+                false,
+            ),
+            (
+                HttpMatcher::custom(TrueMatcher),
+                HttpMatcher::custom(TrueMatcher),
+                HttpMatcher::custom(FalseMatcher),
+                false,
+            ),
+            (
+                HttpMatcher::custom(TrueMatcher),
+                HttpMatcher::custom(FalseMatcher),
+                HttpMatcher::custom(TrueMatcher),
+                false,
+            ),
+            (
+                HttpMatcher::custom(TrueMatcher),
+                HttpMatcher::custom(FalseMatcher),
+                HttpMatcher::custom(FalseMatcher),
+                false,
+            ),
+            (
+                HttpMatcher::custom(FalseMatcher),
+                HttpMatcher::custom(TrueMatcher),
+                HttpMatcher::custom(TrueMatcher),
+                false,
+            ),
+            (
+                HttpMatcher::custom(FalseMatcher),
+                HttpMatcher::custom(TrueMatcher),
+                HttpMatcher::custom(FalseMatcher),
+                false,
+            ),
+            (
+                HttpMatcher::custom(FalseMatcher),
+                HttpMatcher::custom(FalseMatcher),
+                HttpMatcher::custom(TrueMatcher),
+                false,
+            ),
+            (
+                HttpMatcher::custom(FalseMatcher),
+                HttpMatcher::custom(FalseMatcher),
+                HttpMatcher::custom(FalseMatcher),
+                true,
+            ),
+        ];
+
+        for (a, b, c, expected) in cases {
+            let matcher = a.or(b).or(c).negate();
+            let req = Request::builder().body(()).unwrap();
+            assert_eq!(
+                matcher.matches(None, &Context::default(), &req),
+                expected,
+                "({:#?}).matches({:#?})",
+                matcher,
+                req
+            );
+        }
+    }
+
+    #[test]
+    fn test_matcher_or_and_or_and_negation() {
```

Comment:
I would just apply the same technique I propose in my other comment. This way you can easily test it without having to write the cases yourself :) And easier to reason about then these manual http cases. So I would suggest to replace them, but only if you apply a technique such as the ones approached, because manual test cases won't make it that much easier.